language-plutus-core/src/Language/PlutusCore/CBOR.hs

```diff
@@ -1,206 +1,172 @@
-module Language.PlutusCore.CBOR ( encodeProgram
-                                , decodeProgram
-                                , readProgram
-                                , writeProgram
-                                ) where
+{-# LANGUAGE FlexibleInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
```

Comment:
Don't introduce orphan instances.

Reply:
Would you prefer me to move the contents of this file into the various files where the types are defined? I think that gains us nothing except making things a mess.

I agree that orphans are bad, but I think *within the same package* it doesn't matter much.

Reply:
Agree with Michael.

Reply:
I think it's worth elaborating on my reasoning a bit since this is something that may come up again.

Orphans per se aren't bad, what's bad is violating global coherence. Banning orphans is a way of making sure that any time someone has a type T, they also have the instances for T. But as package maintainers we can also ensure that - we can export T and its instances together from an exposed module, and not allow a user to get them separately. Then we've achieved the same goal with some package-level grouping.

This is exactly the case with the instances here (the types and instances are exported together from `PlutusCore`, and not elsewhere), which is why I think it's fine.

(In some cases even violating global coherence isn't that bad - I don't think people using different instances of the pretty printing classes and getting different output is any worse than people using different pretty printing functions and getting different output.)

Reply:
> I agree that orphans are bad, but I think within the same package it doesn't matter much.

Orphans instances are bad in general, and conflicts within the same package are precisely where we can avoid them. 

In fact, Rust bans orphan instances completely. It's not simply a question of different packages.

Reply:
> Orphans instances are bad in general, and conflicts within the same package are precisely where we can avoid them.

Can you elaborate?